src/components/map-controls.js

```diff
@@ -179,213 +109,88 @@ export default class MapControls extends PureComponent {
     if (this.props.isDragging) {
       return config.CURSOR.GRABBING;
     }
-    if (this.state.isHovering) {
+    if (this.props.isHovering) {
       return config.CURSOR.POINTER;
     }
     return config.CURSOR.GRAB;
   }
 
-  _updateViewport(opts) {
-    let viewport = Object.assign({
-      latitude: this.props.latitude,
-      longitude: this.props.longitude,
-      zoom: this.props.zoom,
-      bearing: this.props.bearing,
-      pitch: this.props.pitch,
-      altitude: this.props.altitude,
-      isDragging: this.props.isDragging,
-      startDragLngLat: this.props.startDragLngLat,
-      startBearing: this.props.startBearing,
-      startPitch: this.props.startPitch
-    }, opts);
-
-    viewport = this._applyConstraints(viewport);
-
-    // if (viewport.startDragLngLat) {
-    //   const dragViewport = new FlatMercatorViewport(Object.assign({}, this.props, {
-    //     longitude: viewport.startDragLngLat[0],
-    //     latitude: viewport.startDragLngLat[1]
-    //   }));
-    //   this.setState({dragViewport});
-    // }
-
-    return this.props.onChangeViewport(viewport);
+  _updateViewport(...opts) {
+    const {isDragging} = this.props;
+    return this.props.onChangeViewport(Object.assign({isDragging}, ...opts));
   }
 
-  // Apply any constraints (mathematical or defined by props) to viewport params
-  _applyConstraints(viewport) {
-    // Normalize degrees
-    viewport.longitude = mod(viewport.longitude + 180, 360) - 180;
-    viewport.bearing = mod(viewport.bearing + 180, 360) - 180;
-
-    // Ensure zoom is within specified range
-    const {maxZoom, minZoom} = this.props;
-    viewport.zoom = viewport.zoom > maxZoom ? maxZoom : viewport.zoom;
-    viewport.zoom = viewport.zoom < minZoom ? minZoom : viewport.zoom;
-
-    // Ensure pitch is within specified range
-    const {maxPitch, minPitch} = this.props;
-
-    viewport.pitch = viewport.pitch > maxPitch ? maxPitch : viewport.pitch;
-    viewport.pitch = viewport.pitch < minPitch ? minPitch : viewport.pitch;
-
-    return viewport;
+  _onTouchRotate(opts) {
+    this._onMouseRotate(opts);
   }
 
-  _unproject(pos) {
-    const viewport = new PerspectiveMercatorViewport(this.props);
-    return viewport.unproject(pos, {topLeft: false});
+  _onMouseDown({pos}) {
+    const mapState = new MapState(this.props).panStart({pos}).rotateStart({pos});
+    this._updateViewport(mapState.props, {isDragging: true});
```

Comment:
Should we pass in `mapState` instead so that these functions only need to know about that class and not care about its structure.